csharp/extractor/Semmle.Extraction.CSharp.StubGenerator/StubVisitor.cs

```diff
@@ -0,0 +1,835 @@
+                stubWriter.Write("protected internal ");
+                break;
+            default:
+                stubWriter.Write($"/* TODO: {accessibility} */");
+                break;
+        }
+    }
+
+    private void StubModifiers(ISymbol symbol, bool skipAccessibility = false)
+    {
+        if (symbol.ContainingType is ITypeSymbol containing && containing.TypeKind == TypeKind.Interface)
+            skipAccessibility = true;
+
+        if (symbol is IMethodSymbol method && method.MethodKind == MethodKind.Constructor && symbol.IsStatic)
+            skipAccessibility = true;
+
+        if (!skipAccessibility)
+            StubAccessibility(symbol.DeclaredAccessibility);
+
+        if (symbol.IsAbstract)
+        {
+            if (
+                // exclude interface declarations
+                (symbol is not INamedTypeSymbol type || type.TypeKind != TypeKind.Interface) &&
+                // exclude non-static interface members
+                (symbol.ContainingType is not INamedTypeSymbol containingType || containingType.TypeKind != TypeKind.Interface || symbol.IsStatic))
+            {
+                stubWriter.Write("abstract ");
+            }
+        }
+
+        if (symbol.IsStatic)
+            stubWriter.Write("static ");
+        if (symbol.IsVirtual)
+            stubWriter.Write("virtual ");
+        if (symbol.IsOverride)
+            stubWriter.Write("override ");
+        if (symbol.IsSealed)
+        {
+            if (!(symbol is INamedTypeSymbol type && (type.TypeKind == TypeKind.Enum || type.TypeKind == TypeKind.Delegate || type.TypeKind == TypeKind.Struct)))
+                stubWriter.Write("sealed ");
+        }
+        if (symbol.IsExtern)
+            stubWriter.Write("extern ");
+    }
+
+    public void StubTypedConstant(TypedConstant c)
+    {
+        switch (c.Kind)
+        {
+            case TypedConstantKind.Primitive:
+                if (c.Value is string s)
+                {
+                    stubWriter.Write($"\"{s}\"");
+                }
+                else if (c.Value is char ch)
+                {
+                    stubWriter.Write($"'{ch}'");
+                }
+                else if (c.Value is bool b)
+                {
+                    stubWriter.Write(b ? "true" : "false");
+                }
+                else if (c.Value is int i)
+                {
+                    stubWriter.Write(i);
+                }
+                else if (c.Value is long l)
+                {
+                    stubWriter.Write(l);
+                }
+                else if (c.Value is float f)
+                {
+                    stubWriter.Write(f);
+                }
+                else if (c.Value is double d)
+                {
+                    stubWriter.Write(d);
+                }
+                else
+                {
+                    stubWriter.Write("throw null");
```

Comment:
Does `throw null` work in an attribute argument?

Reply:
Possibly not, but this is mostly dead code for now. We can make it work if/when we need to stub other attributes than `System.FlagsAttribute`.